File: 15.go
package main

import (
	"fmt"
)

type Memory struct {
	first int
	last  int
}

func (m Memory) age() int {
	return m.last - m.first
}

func (m Memory) add(turn int) Memory {
	return Memory{m.last, turn}
}

func main() {
	input := []int{20, 9, 11, 0, 1, 2}
	numbers := make(map[int]Memory, len(input))

	for i, x := range input {
		numbers[x] = Memory{i + 1, i + 1}
	}

	s := input[len(input)-1]
	for i := len(input) + 1; i <= 30000000; i++ {
		s = numbers[s].age()
		if _, ok := numbers[s]; !ok {
			numbers[s] = Memory{i, i}
		} else {
			numbers[s] = numbers[s].add(i)
		}

		if i == 2020 {
			fmt.Println(s)
		}
	}
	fmt.Println(s)

}

File: 15.py
import itertools


class Memory:
    def __init__(self, turn):
        self.first = turn
        self.last = turn

    @property
    def age(self):
        return self.last - self.first

    def add(self, turn):
        self.first, self.last = self.last, turn


if __name__ == "__main__":
    input = [20, 9, 11, 0, 1, 2]

    numbers = {}
    for i, x in enumerate(input):
        numbers[x] = Memory(i + 1)

    s = input[-1]
    for turn in itertools.count(len(input) + 1):
        s = numbers[s].age
        try:
            numbers[s].add(turn)
        except KeyError:
            numbers[s] = Memory(turn)

        if turn == 2020:
            print(s)

        if turn == 30000000:
            print(s)
            break
